### DISCUSSION TITLE
before_prepare / after_prepare: Two phase commit session events

### DISCUSSION THREAD

**Discussion by moi90** [sqlalchemy/sqlalchemy#6384] [Usage Questions]
I would like to hook into two-phase commits using session events. Is there a reason that there is no before_prepare / after_prepare?

My use case is the following: I want to create / delete a directory structure in synchronization with the creation / deletion of a database object. (If an insert fails, no file should be created, and vice-versa. If a delete fails, no files should be deleted, and vice-versa.) Two-phase commits seem to be the right solution to this problem:

```python
Session = sessionmaker(twophase=True)
Base = declarative_base()

class Dataset(Base):
  ...
  def create_files(self):
    ...
  def delete_files(self):
    ...

session = Session()

dataset = Dataset()

session.add(dataset)
session.prepare()
dataset.create_files()
session.commit()

session.delete(dataset)
session.prepare()
dataset.delete_files()
session.commit()
```

Now, I don't want to repeat this dance for the various adds/deletes all over the place. Therefore, I want to automate it: Whenever a `Dataset` instance is `add`ed to the session, its `create_files` should be called automatically after `prepare`. If there was a `"after_prepare"` event, I could use it for that.

Another possibility could be to subclass `Session`:

```python
class PrepareSession(Session):
  def prepare(self):
    super().prepare()
    for inst in self.new:
      inst.prepare("insert")
    for inst in self.deleted:
      inst.prepare("delete")
```

However, I don't know if `Session.new` and `Session.deleted` are still valid after `prepare` (or `flush`) was called.

Is there an obvious solution that I'm overlooking?

**Comment by zzzeek** [comment]
are you going to actually use 2PC for real on your database backend?  because that's a heavy-handed approach here. PostgreSQL only allows a fixed number of these at a time and if they are not closed they are permanent until you go in and delete them.    I believe that calling session.prepare() explicitly right now is checking that 2pc is in use.

that said, the "prepare" step is still used internally even if 2pc isn't present so if we added an "after_prepare(soft=True)" kind of event , where "soft" means "we didnt actually do a 2pc PREPARE, we just did the final flush before commit", that would be useful for this, so I would suggest proposing that as a feature.

**Reply by zzzeek** [reply]
I would propose session.prepare(soft_ok=True) as well which means, "just do the internal prepare and it's OK if it's not real 2pc"

**Reply by moi90** [reply]
Yes, I'm using PostgreSQL.
I would need 2PC only for this one table. Does it bring a large overhead if I otherwise directly use `session.commit()` without a prior `session.prepare()`?

I like the "soft prepare" proposal, but I currently don't have the time to advocate for it.

If I solve my problem without 2PC, I could create the directory after flush (and delete on rollback) for an added object -- and delete the directory after commit for a deleted object. The downside is that if the removal of the directory fails, nothing is pointing to it any more which necessitates periodical checks for orphaned directories.

**Reply by zzzeek** [reply]
> Yes, I'm using PostgreSQL.
> I would need 2PC only for this one table. Does it bring a large overhead if I otherwise directly use `session.commit()` without a prior `session.prepare()`?

the .commit() method does the .prepare() step internally, that is, within the ORM "session" the concept of "prepare" is part of the architecture whether or not real 2pc transactions are used.   prepare() indicates, lets do our final flush, and some other bookkeeping, and then get ready for commit().   so i dont see any "overhead" related variance here, only if you have real 2pc in use, then the database itself probably has some more overhead.

> 
> I like the "soft prepare" proposal, but I currently don't have the time to advocate for it.
> 
> If I solve my problem without 2PC, I could create the directory after flush (and delete on rollback) for an added object -- and delete the directory after commit for a deleted object. The downside is that if the removal of the directory fails, nothing is pointing to it any more which necessitates periodical checks for orphaned directories.

yes usually people do this kind of thing as part of "after_flush()", which has an event, and you can also rollback the transaction within after_flush() too.    buf after_flush() only triggers if there's actually something to flush so this wouldn't be a hook that works as a "prepare" step here.

at the moment the closest thing to a "prepare" event is to do the before_commit() event and flush the session inside of it, which should be fine:

```python
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)

e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

@event.listens_for(s, "before_commit")
def _emulate_prepare_event(session):
    session.flush()

    # do prepare stuff
    print("prepare step")

s.add(A(data='x'))

s.commit()
```


so if your prepare fails you can still raise an exception that will prevent the transaction from committing.
